cmd/osm-controller/osm-controller_test.go

```diff
@@ -2,74 +2,211 @@ package main
+
+	cfg := configurator.NewConfigurator(kubeClient, stop, osmNamespace, osmConfigMapName)
+	errCh := make(chan error, 1)
+
+	fakeDebugServer := FakeDebugServer{0, nil}
+	fakeDebugServerGetErr := FakeDebugServer{0, errors.Errorf("Debug server error")}
+
+	testCases := []struct {
+		name                       string
+		initialDebugServerEnabled  bool
+		changeDebugServerEnabledTo bool
+		c                          controller
+		expectedStopCount          int
+		expectedStopErr            bool
+		expectedDebugServerRunning bool
+	}{
+		{
+			name:                       "debug server is already on, do nothing",
+			initialDebugServerEnabled:  true,
+			changeDebugServerEnabledTo: true,
+			c:                          controller{debugServerRunning: true, debugComponents: mockDebugConfig, debugServer: nil},
+			expectedStopCount:          0,
+			expectedStopErr:            false,
+			expectedDebugServerRunning: true,
+		},
+		{
+			name:                       "debug server is already off, do nothing",
+			initialDebugServerEnabled:  false,
+			changeDebugServerEnabledTo: false,
+			c:                          controller{debugServerRunning: false, debugComponents: mockDebugConfig, debugServer: nil},
+			expectedStopCount:          0,
+			expectedStopErr:            false,
+			expectedDebugServerRunning: false,
+		},
+		{
+			name:                       "turn on debug server",
+			initialDebugServerEnabled:  false,
+			changeDebugServerEnabledTo: true,
+			c:                          controller{debugServerRunning: false, debugComponents: mockDebugConfig, debugServer: nil},
+			expectedStopCount:          0,
+			expectedStopErr:            false,
+			expectedDebugServerRunning: true,
+		},
+		{
+			name:                       "turn off debug server",
+			initialDebugServerEnabled:  true,
+			changeDebugServerEnabledTo: false,
+			c:                          controller{debugServerRunning: true, debugComponents: mockDebugConfig, debugServer: &fakeDebugServer},
+			expectedStopCount:          1,
+			expectedStopErr:            false,
+			expectedDebugServerRunning: false,
+		},
+		{
+			name:                       "error when turning off debug server",
+			initialDebugServerEnabled:  true,
+			changeDebugServerEnabledTo: false,
+			c:                          controller{debugServerRunning: true, debugComponents: mockDebugConfig, debugServer: &fakeDebugServerGetErr},
+			expectedStopCount:          1,
+			expectedStopErr:            true,
+			expectedDebugServerRunning: false,
+		},
+	}
+	firstTest := true
+
+	for _, tests := range testCases {
+		t.Run(fmt.Sprintf("Test: %s", tests.name), func(t *testing.T) {
+			defaultConfigMap := map[string]string{
+				"enabled_debug_server": strconv.FormatBool(tests.initialDebugServerEnabled),
+			}
+			configMap := v1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: osmNamespace,
+					Name:      osmConfigMapName,
+				},
+				Data: defaultConfigMap,
+			}
+			if firstTest {
```

Comment:
setup should happen before running the test cases and cleanup should happen after asserting results. you want to make sure order doesn't matter when running these kinds of tests. Every test should be run in an isolates setting and created just for that test.